src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs

```diff
@@ -34,36 +34,57 @@ private static Half ReadCore(ref Utf8JsonReader reader)
         {
             Half result;
 
-            if (reader.ValueLength > MaxEscapedFormatLength)
+            byte[]? rentedByteBuffer = null;
+            char[]? rentedCharBuffer = null;
+            int bufferLength = reader.ValueLength;
+
+            try
             {
-                ThrowHelper.ThrowFormatException(NumericType.Half);
-            }
+                Span<byte> byteBuffer = bufferLength <= JsonConstants.StackallocByteThreshold
+                    ? stackalloc byte[JsonConstants.StackallocCharThreshold]
+                    : (rentedByteBuffer = ArrayPool<byte>.Shared.Rent(bufferLength));
 
-            Span<byte> buffer = stackalloc byte[MaxFormatLength];
-            int written = reader.CopyString(buffer);
-            buffer = buffer.Slice(0, written);
+                int written = reader.CopyValue(byteBuffer);
+                byteBuffer = byteBuffer.Slice(0, written);
 
-            if (reader.TokenType is JsonTokenType.String or JsonTokenType.PropertyName)
-            {
-                if (JsonReaderHelper.TryGetFloatingPointConstant(buffer, out result))
+                if (reader.TokenType is JsonTokenType.String or JsonTokenType.PropertyName)
                 {
-                    return result;
+                    if (JsonReaderHelper.TryGetFloatingPointConstant(byteBuffer, out result))
+                    {
+                        return result;
+                    }
                 }
-            }
 
 #if NET8_0_OR_GREATER
-            bool success = Half.TryParse(buffer, out result);
+                bool success = Half.TryParse(byteBuffer, out result);
 #else
-            Span<char> charBuffer = stackalloc char[MaxFormatLength];
-            written = JsonReaderHelper.TranscodeHelper(buffer, charBuffer);
-            bool success = Half.TryParse(charBuffer.Slice(0, written), out result);
+                // Half.TryParse(ROS<byte>) is not available on .NET 7, only Half.TryParse(ROS<char>);
+                // we need to transcode here instead of letting CopyValue do it for us because
+                // TryGetFloatingPointConstant only accepts ROS<byte>.
+
+                Span<char> charBuffer = stackalloc char[MaxFormatLength];
+                written = JsonReaderHelper.TranscodeHelper(byteBuffer, charBuffer);
+                bool success = Half.TryParse(charBuffer.Slice(0, written), out result);
 #endif
-            if (!success)
-            {
-                ThrowHelper.ThrowFormatException(NumericType.Half);
+                if (!success)
+                {
+                    ThrowHelper.ThrowFormatException(NumericType.Half);
+                }
+
+                return result;
             }
+            finally
```

Comment:
We generally don't use try/finally to guard against exceptions in code that uses rented buffers. Not returning a buffer because of an exception is not a big problem all things considered (the problems start if a buffer gets used after being returns or gets returned more than once). You should still move the returning logic above the `if (!success)` statement above though.